0018-publish-prompt.md

```diff
@@ -0,0 +1,67 @@
+# Publish prompt
+
+## Summary
+
+Publishing a package should prompt a confirmation prompt, allowing users to validate their package info before uploading their package tarball.
+
+## Motivation
+
+Breaking the prompt into a two-step operation allows for a validation of contents prior to uploading the package but it also enable new possibilities such as printing warnings promotting 2FA adoption, etc.
+
+## Detailed Explanation and Rationale
+
+`npm publish` should ask for a confirmation prompt prior to uploading the package tarball.
+
+Currently "publish" has a very nice behavior for users that have 2FA enabled, one in which it stops the publish process in order to ask for the OTP code, allowing for a review of the file contents prior to uploading the tarball file.
+
+It would be very nice to expand that "review" 
+
+## Alternatives
+
+Not change it, just keep the current behavior?
+
+## Implementation
+
+This is a breaking change from the current `npm publish` behavior, it would prompt the user for confirmation:
```

Comment:
perhaps a good first step would be, provide an option to enable the prompt, and then change the default in a future major?

Reply:
We could just make the default be `--yes` when `process.stdout.isTTY === false`.  Then it'll keep on working for build scripts and CI, but prompt everyone on the CLI.

Reply:
Do you have plans to align with other package managers?  Would be great to have a consistent behavior, I think this is a great idea.

Reply:
I agree, I think we would do this as well in Yarn. cc @zkochan

Reply:
Seems fine for single-package scenario but I am not sure this would work during a publish in a monorepo. I doubt that someone would review the content of every published package if dozens are published.

Reply:
What if it would only show the prompt if the list of files that are published differs from the list of files in the previously published version?

Reply:
Not all monorepos publish multiple packages at once; in my experience many use independent versioning at this point.

Reply:
Printing a file diff if possible could be interesting 🤔

Something else to consider is that workflows publishing multiple packages (which do exist, for example this is what we do on Yarn, and I think Babel and Jest do the same) might be more likely to just use `--yes` (which would be handy, but would kinda negate this RFC altogether).

Reply:
multiple packages is common; "bumping all package versions at once even if they haven't changed, to keep all versions the same" is the original lerna behavior that babel, and many others, have thankfully abandoned :-)

Reply:
Clarifying Lerna behavior for future reference:

> "bumping all package versions at once even if they haven't changed, to keep all versions the same"

This isn't actually what Lerna does in the default "fixed" versioning scheme. It applies a single version to any package (and its local dependents) that has changed since the previous release (annotated git tag).

> Not all monorepos publish multiple packages at once; in my experience many use independent versioning at this point.

"Independent" versioning doesn't change anything about how Lerna detects changes between releases or bumps dependents of changed packages, it merely allows one to choose versions per-package instead of globally.